### PR TITLE
feat: ExternalOpsTracker

### PR DESCRIPTION
**Pull request by devsnek**
Allows an embedder to track "external" operations which should still keep the event loop alive. Will be used to ref napi threadsafe functions.

### CODE REVIEW

**Review by bartlomieju — COMMENTED** [review]
@devsnek could you run some of the benchmarks we have on `main` branch and in this branch to see what's the perf delta?

**Comment by devsnek** [comment]
benchmarks:

<details>
<summary>main</summary>

```
     Running benches/ops/async.rs (target/release/deps/ops_async-05d79c7f29673bb8)

running 17 tests
test baseline                             ... bench:       2,221 ns/iter (+/- 75)
test baseline_op_promise                  ... bench:     105,403 ns/iter (+/- 1,755)
test baseline_op_promise_resolver         ... bench:     105,905 ns/iter (+/- 5,846)
test baseline_promise                     ... bench:      37,031 ns/iter (+/- 2,565)
test baseline_promise_with_resolver       ... bench:      53,008 ns/iter (+/- 2,238)
test baseline_sync                        ... bench:      40,539 ns/iter (+/- 9,177)
test bench_op_async_void                  ... bench:      58,495 ns/iter (+/- 19,758)
test bench_op_async_void_deferred         ... bench:     819,373 ns/iter (+/- 59,497)
test bench_op_async_void_deferred_nofast  ... bench:     822,485 ns/iter (+/- 31,746)
test bench_op_async_void_deferred_return  ... bench:     821,449 ns/iter (+/- 36,473)
test bench_op_async_void_lazy             ... bench:     765,202 ns/iter (+/- 43,588)
test bench_op_async_void_lazy_nofast      ... bench:     788,607 ns/iter (+/- 35,933)
test bench_op_async_yield                 ... bench:     853,732 ns/iter (+/- 48,133)
test bench_op_async_yield_deferred        ... bench:     826,599 ns/iter (+/- 25,274)
test bench_op_async_yield_deferred_nofast ... bench:     834,053 ns/iter (+/- 25,083)
test bench_op_async_yield_lazy            ... bench:     932,720 ns/iter (+/- 31,789)
test bench_op_async_yield_lazy_nofast     ... bench:     954,030 ns/iter (+/- 17,526)

test result: ok. 0 passed; 0 failed; 0 ignored; 17 measured

     Running benches/ops/sync.rs (target/release/deps/ops_sync-320bc4bdcdc1b3f4)

running 33 tests
test baseline                              ... bench:         249 ns/iter (+/- 6)
test bench_op_arraybuffer                  ... bench:       5,898 ns/iter (+/- 133)
test bench_op_bigint                       ... bench:       3,292 ns/iter (+/- 85)
test bench_op_bigint_return                ... bench:       3,748 ns/iter (+/- 117)
test bench_op_buffer                       ... bench:       2,924 ns/iter (+/- 1,185)
test bench_op_buffer_jsbuffer              ... bench:      39,041 ns/iter (+/- 2,267)
test bench_op_buffer_nofast                ... bench:      42,796 ns/iter (+/- 942)
test bench_op_external                     ... bench:       3,317 ns/iter (+/- 82)
test bench_op_external_nofast              ... bench:       7,975 ns/iter (+/- 179)
test bench_op_option_u32                   ... bench:       7,542 ns/iter (+/- 67)
test bench_op_string                       ... bench:       9,164 ns/iter (+/- 762)
test bench_op_string_bytestring            ... bench:      44,173 ns/iter (+/- 1,070)
test bench_op_string_large_1000            ... bench:     120,444 ns/iter (+/- 3,887)
test bench_op_string_large_1000000         ... bench:     477,616 ns/iter (+/- 15,331)
test bench_op_string_large_utf8_1000       ... bench:   1,485,077 ns/iter (+/- 82,200)
test bench_op_string_large_utf8_1000000    ... bench:   7,220,498 ns/iter (+/- 240,440)
test bench_op_string_onebyte               ... bench:       3,285 ns/iter (+/- 260)
test bench_op_string_onebyte_large_1000    ... bench:      32,365 ns/iter (+/- 554)
test bench_op_string_onebyte_large_1000000 ... bench:      81,862 ns/iter (+/- 1,461)
test bench_op_string_option_u32            ... bench:      27,342 ns/iter (+/- 486)
test bench_op_u32                          ... bench:       3,268 ns/iter (+/- 101)
test bench_op_v8_global                    ... bench:      23,158 ns/iter (+/- 389)
test bench_op_v8_global_scope              ... bench:      26,088 ns/iter (+/- 1,123)
test bench_op_v8_isolate_nofast            ... bench:       7,507 ns/iter (+/- 129)
test bench_op_v8_local                     ... bench:       3,493 ns/iter (+/- 76)
test bench_op_v8_local_nofast              ... bench:       7,550 ns/iter (+/- 183)
test bench_op_v8_local_scope               ... bench:      12,034 ns/iter (+/- 280)
test bench_op_v8_scope                     ... bench:       9,475 ns/iter (+/- 476)
test bench_op_void                         ... bench:       3,328 ns/iter (+/- 93)
test bench_op_void_2x                      ... bench:       3,122 ns/iter (+/- 41)
test bench_op_void_metrics                 ... bench:       4,943 ns/iter (+/- 234)
test bench_op_void_nofast                  ... bench:       6,739 ns/iter (+/- 54)
test bench_op_void_nofast_metrics          ... bench:       7,909 ns/iter (+/- 357)

test result: ok. 0 passed; 0 failed; 0 ignored; 33 measured
```

</details>

<details>
<summary>this pr</summary>

```
     Running benches/ops/async.rs (target/release/deps/ops_async-05d79c7f29673bb8)

running 17 tests
test baseline                             ... bench:       2,214 ns/iter (+/- 233)
test baseline_op_promise                  ... bench:     103,452 ns/iter (+/- 4,184)
test baseline_op_promise_resolver         ... bench:     104,829 ns/iter (+/- 3,090)
test baseline_promise                     ... bench:      36,873 ns/iter (+/- 2,885)
test baseline_promise_with_resolver       ... bench:      52,704 ns/iter (+/- 1,360)
test baseline_sync                        ... bench:      40,585 ns/iter (+/- 2,970)
test bench_op_async_void                  ... bench:      57,657 ns/iter (+/- 1,655)
test bench_op_async_void_deferred         ... bench:     813,552 ns/iter (+/- 28,204)
test bench_op_async_void_deferred_nofast  ... bench:     834,604 ns/iter (+/- 80,972)
test bench_op_async_void_deferred_return  ... bench:     830,107 ns/iter (+/- 28,318)
test bench_op_async_void_lazy             ... bench:     777,467 ns/iter (+/- 24,162)
test bench_op_async_void_lazy_nofast      ... bench:     805,703 ns/iter (+/- 48,009)
test bench_op_async_yield                 ... bench:     837,649 ns/iter (+/- 31,705)
test bench_op_async_yield_deferred        ... bench:     833,549 ns/iter (+/- 29,541)
test bench_op_async_yield_deferred_nofast ... bench:     848,649 ns/iter (+/- 21,555)
test bench_op_async_yield_lazy            ... bench:     970,860 ns/iter (+/- 42,998)
test bench_op_async_yield_lazy_nofast     ... bench:     968,059 ns/iter (+/- 18,792)

test result: ok. 0 passed; 0 failed; 0 ignored; 17 measured

     Running benches/ops/sync.rs (target/release/deps/ops_sync-320bc4bdcdc1b3f4)

running 33 tests
test baseline                              ... bench:         249 ns/iter (+/- 1)
test bench_op_arraybuffer                  ... bench:       5,783 ns/iter (+/- 150)
test bench_op_bigint                       ... bench:       3,320 ns/iter (+/- 108)
test bench_op_bigint_return                ... bench:       3,761 ns/iter (+/- 91)
test bench_op_buffer                       ... bench:       2,906 ns/iter (+/- 750)
test bench_op_buffer_jsbuffer              ... bench:      38,216 ns/iter (+/- 828)
test bench_op_buffer_nofast                ... bench:      42,907 ns/iter (+/- 1,540)
test bench_op_external                     ... bench:       3,324 ns/iter (+/- 28)
test bench_op_external_nofast              ... bench:       7,770 ns/iter (+/- 271)
test bench_op_option_u32                   ... bench:       6,985 ns/iter (+/- 79)
test bench_op_string                       ... bench:       8,989 ns/iter (+/- 163)
test bench_op_string_bytestring            ... bench:      43,500 ns/iter (+/- 1,193)
test bench_op_string_large_1000            ... bench:     115,908 ns/iter (+/- 2,591)
test bench_op_string_large_1000000         ... bench:     479,775 ns/iter (+/- 22,607)
test bench_op_string_large_utf8_1000       ... bench:   1,486,399 ns/iter (+/- 36,938)
test bench_op_string_large_utf8_1000000    ... bench:   7,224,838 ns/iter (+/- 19,381)
test bench_op_string_onebyte               ... bench:       3,263 ns/iter (+/- 55)
test bench_op_string_onebyte_large_1000    ... bench:      31,298 ns/iter (+/- 673)
test bench_op_string_onebyte_large_1000000 ... bench:      82,440 ns/iter (+/- 7,375)
test bench_op_string_option_u32            ... bench:      27,266 ns/iter (+/- 667)
test bench_op_u32                          ... bench:       3,256 ns/iter (+/- 194)
test bench_op_v8_global                    ... bench:      21,746 ns/iter (+/- 1,806)
test bench_op_v8_global_scope              ... bench:      25,823 ns/iter (+/- 603)
test bench_op_v8_isolate_nofast            ... bench:       7,476 ns/iter (+/- 148)
test bench_op_v8_local                     ... bench:       3,501 ns/iter (+/- 211)
test bench_op_v8_local_nofast              ... bench:       7,512 ns/iter (+/- 134)
test bench_op_v8_local_scope               ... bench:      11,235 ns/iter (+/- 270)
test bench_op_v8_scope                     ... bench:       9,457 ns/iter (+/- 526)
test bench_op_void                         ... bench:       3,295 ns/iter (+/- 75)
test bench_op_void_2x                      ... bench:       3,118 ns/iter (+/- 34)
test bench_op_void_metrics                 ... bench:       4,938 ns/iter (+/- 143)
test bench_op_void_nofast                  ... bench:       6,724 ns/iter (+/- 80)
test bench_op_void_nofast_metrics          ... bench:       7,789 ns/iter (+/- 109)

test result: ok. 0 passed; 0 failed; 0 ignored; 33 measured
```

</details>

tldr: no change

**Review by bartlomieju — APPROVED** [review]
Okay, let's try this one. We might have to revisit if we land this in Deno and we'll see a dent in the benchmarks (eg. HTTP server benches)
